File: api/src/main/java/work/beltran/discogsbrowser/api/DiscogsService.java
package work.beltran.discogsbrowser.api;

import retrofit2.http.DELETE;
import retrofit2.http.GET;
import retrofit2.http.POST;
import retrofit2.http.Path;
import retrofit2.http.Query;
import work.beltran.discogsbrowser.api.model.SearchResults;
import work.beltran.discogsbrowser.api.model.UserCollection;
import work.beltran.discogsbrowser.api.model.UserIdentity;
import work.beltran.discogsbrowser.api.model.UserProfile;
import work.beltran.discogsbrowser.api.model.UserWanted;

/**
 * Created by Miquel Beltran on 22.04.16
 * More on http://beltran.work
 */
public interface DiscogsService {
    // display https://api.discogs.com/users/mike513/collection/folders/0/releases
    @GET("users/{user}/collection/folders/0/releases")
    rx.Observable<UserCollection> listRecords(@Path("user") String user,
                                              @Query("page") int page);

    @GET("users/{username}/collection/releases/{release_id}")
    rx.Observable<UserCollection> getRecordInCollection(@Path("username") String user,
                                                        @Path("release_id") int releaseId);

    @POST("users/{username}/collection/folders/1/releases/{release_id}")
    rx.Observable<Void> addToCollection(@Path("username") String user,
                                        @Path("release_id") int releaseId);

    @DELETE("users/{username}/collection/folders/1/releases/{release_id}/instances/{instance_id}")
    rx.Observable<Void> removeFromCollection(@Path("username") String user,
                                             @Path("release_id") int releaseId,
                                             @Path("instance_id") int instanceId);

    @GET("users/{username}")
    rx.Observable<UserProfile> getUserProfile(@Path("username") String username);

    @GET("/users/{username}/wants")
    rx.Observable<UserWanted> getWantedList(@Path("username") String username,
                                            @Query("page") int page);

    @GET("database/search")
    rx.Observable<SearchResults> search(@Query("q") String query,
                                        @Query("type") String type,
                                        @Query("format") String format,
                                        @Query("barcode") String barcode);

    @GET("oauth/identity")
    rx.Observable<UserIdentity> getUserIdentity();
}

File: api/src/main/java/work/beltran/discogsbrowser/api/DiscogsService.kt
package work.beltran.discogsbrowser.api

import com.jakewharton.retrofit2.adapter.kotlin.coroutines.CoroutineCallAdapterFactory
import com.squareup.moshi.Moshi
import com.squareup.moshi.kotlin.reflect.KotlinJsonAdapterFactory
import kotlinx.coroutines.Deferred
import retrofit2.Response
import retrofit2.Retrofit
import retrofit2.converter.moshi.MoshiConverterFactory
import retrofit2.http.GET
import retrofit2.http.Path
import retrofit2.http.Query

interface DiscogsService {

    @GET("users/{username}/collection/folders/{folder_id}/releases")
    fun getCollectionItemsByFolder(
        @Path("username") username: String,
        @Path("folder_id") folderId: String,
        @Query("page") page: Int,
        @Query("per_page") perPage: Int
    ) : Deferred<Response<CollectionItemsByFolderResponse>>
}

fun provideService(): DiscogsService {

    val moshi = Moshi.Builder()
        .add(KotlinJsonAdapterFactory())
        .build()

    return Retrofit
        .Builder()
        .baseUrl("https://api.discogs.com")
        .addConverterFactory(MoshiConverterFactory.create(moshi))
        .addCallAdapterFactory(CoroutineCallAdapterFactory())
        .build()
        .create(DiscogsService::class.java)
}
